### PR TITLE
SB first Merge Request

### PR DESCRIPTION
**Pull request by GilLifshits**

### CODE REVIEW

File: config.py [hunk 1]
```diff
@@ -5,43 +5,35 @@
 class Config:
```
Comment (GilLifshits) [inline config.py]: Go over the config params, If there is no use in some params -> delete them.
If you use them, but do not pull the value from config file, then use the config file.

File: main.py [hunk 2]
```diff
@@ -1,38 +1,38 @@
 import config
 from utils.experiment import Experiment
 from utils.training_loop import *
-
-
+from training_loop import training_loop_ariel_step
```
Comment (GilLifshits) [inline main.py]: change name from training_loop_ariel_step, to other name, pay attention that the training loop is per trajectory/epoch and not step.

File: main.py [hunk 3]
```diff
@@ -1,38 +1,38 @@
...
+from datetime import datetime
+exp_id='Car_2_Random_Side'
```
Comment (GilLifshits) [inline main.py]: Use experiment Id from config

File: main.py [hunk 4]
```diff
@@ -1,38 +1,38 @@
...
+path = 'experiments/' + datetime.now().strftime("%d_%m_%Y-%H_%M_%S") +exp_id
```
Comment (GilLifshits) [inline main.py]: Define path in config

File: utils/NN_utils.py [hunk 5]
```diff
@@ -118,27 +98,24 @@ def create_network_copy(network):
...
             raise FileNotFoundError(f"Weight directory {weight_directory} does not exist.")
-
-        network.load_weights(weight_directory)
+        network=PPO.load(weight_directory)
         print(f"Weights from: {weight_directory} were loaded successfully.")
         return network
 
+    def save_network_weights(self, network):
+        # Create the directory if it doesn't exist
+        save_dir = f"experiments/{self.config.EXPERIMENT_ID}/weights"
```
Comment (GilLifshits) [inline utils/NN_utils.py]: save_dir can be defined in config

File: utils/NN_utils.py [hunk 6]
```diff
@@ -118,27 +98,24 @@ def create_network_copy(network):
...
+        if not os.path.exists(save_dir):
+            os.makedirs(save_dir)
+        # Save the entire PPO model as a .zip file
+        save_path = f"{save_dir}/{self.config.WEIGHTS_TO_SAVE_NAME}.zip"
```
Comment (GilLifshits) [inline utils/NN_utils.py:114]: save_path can be defined in config

File: utils/airsim_manager.py [hunk 7]
```diff
@@ -133,25 +128,17 @@ def get_cars_distance(self):
             np.array([[car2_position_and_speed["x"], car2_position_and_speed["y"]]])))
         return dist_c1_c2
 
-    def get_car1_state(self, logger):
+    def get_car1_state(self, logger=None):
```
Comment (GilLifshits) [inline utils/airsim_manager.py:145]: car state should be as defined in notion, am I missing something? :)

File: utils/airsim_manager.py [hunk 8]
```diff
@@ -133,25 +128,17 @@ def get_cars_distance(self):
...
-            # self.get_cars_distance()
         ])
 
-        # if self.config.LOG_CAR_STATES:
-        #     logger.log_state(car1_state, self.config.CAR1_NAME)
+        if logger is not None and self.config.LOG_CAR_STATES:
+            logger.log_state(car1_state, self.config.CAR1_NAME)
 
-        return car1_state
+        return car1_state  # מחזיר מערך NumPy
```
Comment (GilLifshits) [inline utils/airsim_manager.py]: comments should be in English

File: utils/training_loop.py [hunk 9]
```diff
@@ -1,270 +1,104 @@
...
+import pandas as pd
+from config import Config
+import random
+
+def training_loop_ariel_step(config, path):
+    all_rewards = []
+    new_logger = configure(path, ["stdout", "csv", "tensorboard"])
+    env = DummyVecEnv([lambda: AirSimGymEnv(config, AirsimManager(config))])
+    model = PPO('MlpPolicy', env, verbose=1, learning_rate=0.0001, n_steps=160, batch_size=160)
+    #model = PPO('MlpPolicy', env, verbose=1)
```
Comment (GilLifshits) [inline utils/training_loop.py]: remove this comment

File: utils/training_loop.py [hunk 10]
```diff
@@ -1,270 +1,104 @@
...
+from stable_baselines3.common.logger import configure
+import pandas as pd
+from config import Config
+import random
+
+def training_loop_ariel_step(config, path):
+    all_rewards = []
+    new_logger = configure(path, ["stdout", "csv", "tensorboard"])
+    env = DummyVecEnv([lambda: AirSimGymEnv(config, AirsimManager(config))])
+    model = PPO('MlpPolicy', env, verbose=1, learning_rate=0.0001, n_steps=160, batch_size=160)
```
Comment (GilLifshits) [inline utils/training_loop.py]: consider moving the variables to config or when defining an experiment

File: utils/training_loop.py [hunk 11]
```diff
@@ -1,270 +1,104 @@
...
+    #model = PPO('MlpPolicy', env, verbose=1)
+    model.set_logger(new_logger)
```
Comment (GilLifshits) [inline utils/training_loop.py]: new logger? why not just logger?

File: utils/training_loop.py [hunk 12]
```diff
@@ -1,270 +1,104 @@
...
+
+    def train_model():
+        total_timesteps = 160
```
Comment (GilLifshits) [inline utils/training_loop.py]: 160 is constant - can be moved to config/experiment definition

File: utils/training_loop.py [hunk 13]
```diff
@@ -1,270 +1,104 @@
...
+    total_steps = 0
 
-    # Main loop for each episode
     for episode in range(config.MAX_EPISODES):
-
-        # choose constant action for car2 for whole trajectory
-        if np.random.randint(2) == 1:
-            config.CAR2_CONSTANT_ACTION = 0
+        print(f"@ Episode {episode + 1} @")
+        car2_side = random.choice(["left", "right"])
```
Comment (GilLifshits) [inline utils/training_loop.py]: consider moving this section into a function in airsim manager, and also consider make hard coded values (90/[0,-30]) to config/class definition.

File: utils/training_loop.py [hunk 14]
```diff
@@ -1,270 +1,104 @@
...
-        print(f"@ Episode {episode} @")
+            config.CAR2_INITIAL_POSITION = [0, 30]
+            config.CAR2_INITIAL_YAW = 270
 
-        rl.current_trajectory = []
-
-        # Inner loop for each step in an episode
-        for step in range(config.MAX_STEPS):
-            steps_counter += 1
+        print(f"Car 2 starts from {car2_side} with position {config.CAR2_INITIAL_POSITION} and yaw {config.CAR2_INITIAL_YAW}")
```
Comment (GilLifshits) [inline utils/training_loop.py]: Good debugging! well done

File: utils/training_loop.py [hunk 15]
```diff
@@ -1,270 +1,104 @@
...
+        obs = env.reset()
+        done = False
+        episode_sum_of_rewards = 0
+        episode_counter += 1
 
-            # Update the sum of rewards
+        while not done:
+            # print(f"Before step: Done={done}, State={obs}")
+            if not config.ONLY_INFERENCE:
+                if total_steps > 2500:
```
Comment (GilLifshits) [inline utils/training_loop.py]: 2500 should move to config. (+ add to notion conclusion about how to do exploration/explotation in PPO, I do not know if this is the best way)

File: utils/training_loop.py [hunk 16]
```diff
@@ -1,270 +1,104 @@
...
-                # Log the episode's sum of rewards
-                logger.log_scaler("episode_sum_of_rewards", episode_counter, episode_sum_of_rewards)
-
-                # Update collision count and break the loop
-                if collision_occurred:
+            #print('Step:', steps_counter)
+            #print('Reward:', reward)
+            if done:
+                if not config.ONLY_INFERENCE:
+                    env.envs[0].pause_simulation()
```
Comment (GilLifshits) [inline utils/training_loop.py]: Nice solution!

File: utils/training_loop.py [hunk 17]
```diff
@@ -1,270 +1,104 @@
...
-
+        print(f"Episode {episode_counter} finished with reward: {episode_sum_of_rewards}")
+        all_rewards.append(episode_sum_of_rewards)
+        steps_counter = 0
+    model.save(path + '/model')
+    print('Model saved')
+    print("Total collisions:", collision_counter)
+    loss_data = pd.read_csv(f"{path}/progress.csv")
+    losses = loss_data["train/value_loss"]
+    plt.figure(figsize=(12, 6))
```
Comment (GilLifshits) [inline utils/training_loop.py]: move this code to a function (maybe add plotting utils...

**Review by GilLifshits — COMMENTED** [review]
Nice work!

File: config.py (hunk 1 above)
Comment (arielharush96) [reply config.py]: LOG_Q_VALUE ,'LOG_CAR_STATES', 'EPOCHS' , deleted

File: main.py (hunk 2 above)
Comment (arielharush96) [reply main.py]: changed to model_training

File: utils/airsim_manager.py (hunk 8 above)
Comment (arielharush96) [reply utils/airsim_manager.py]: Weird, at my PC the comment didn't appear..

File: main.py (hunk 4 above)
Comment (arielharush96) [reply main.py]: fixed

File: utils/NN_utils.py (hunk 5 above)
Comment (arielharush96) [reply utils/NN_utils.py]: fixed

File: main.py (hunk 3 above)
Comment (arielharush96) [reply main.py]: fixed

File: utils/training_loop.py (hunk 9 above)
Comment (arielharush96) [reply utils/training_loop.py]: fixed

File: utils/training_loop.py (hunk 10 above)
Comment (arielharush96) [reply utils/training_loop.py]: fixed

File: utils/training_loop.py (hunk 11 above)
Comment (arielharush96) [reply utils/training_loop.py]: just a name :-D

File: utils/training_loop.py (hunk 12 above)
Comment (arielharush96) [reply utils/training_loop.py]: func deleted

File: utils/training_loop.py (hunk 13 above)
Comment (arielharush96) [reply utils/training_loop.py]: fixed

File: utils/training_loop.py (hunk 14 above)
Comment (arielharush96) [reply utils/training_loop.py]: Toda :-)

File: utils/training_loop.py (hunk 16 above)
Comment (arielharush96) [reply utils/training_loop.py]: Toda :-)

File: utils/training_loop.py (hunk 15 above)
Comment (arielharush96) [reply utils/training_loop.py]: fixed

File: utils/training_loop.py (hunk 17 above)
Comment (arielharush96) [reply utils/training_loop.py]: fixed

File: utils/airsim_manager.py (hunk 7 above)
Comment (arielharush96) [reply utils/airsim_manager.py:145]: fixed

**Pull request merged**